### ISSUE TITLE
Allow Mixins to work from a Java Agent

### ISSUE THREAD

**Issue by me4502** [SpongePowered/Mixin#111]
Allowing mixins to work from a Java Agent could allow some cool stuff, like injecting them into already loaded classes using something like this, https://github.com/Xyene/ASM-Late-Bind-Agent.

The only issue I see is that AFAIK the Java Agent is incapable of performing a fair few actions, such as adding new methods, or removing methods. Due to this, only Overwrite, Inject, etc, would really work.

**Comment by gabizou** [comment]
> Java Agent is incapable of performing a fair few actions, such as adding new methods, or removing methods.

This basically would prevent anything being added **except** applying `Overwrite`s. Injections are handled where there's a new method added and existing method modified to direct to the newly added method.

**Comment by Mumfrey** [comment]
Can you elaborate on the use case? It doesn't seem that compelling to me but maybe I'm missing something?

> Allowing mixins to work from a Java Agent could allow some cool stuff, like injecting them into already loaded classes using something like this, https://github.com/Xyene/ASM-Late-Bind-Agent.

But why? Under what circumstances would you need to inject into an already loaded class? And how would it actually help?

> The only issue I see is that AFAIK the Java Agent is incapable of performing a fair few actions, such as adding new methods, or removing methods. Due to this, only Overwrite, Inject, etc, would really work.

This is true, so it's not like you could use it as a quick way of starting to write a "real" mixin whilst the game is live (assuming all the other hurdles could be overcome), you could only write a kind of limited mixin, and even then it would only be useful for the first run, since afterwards you know the mixin exists and it can be loaded normally.

It's not that I don't think it's possible, but I just kind of don't see the point *beyond* "it would be cool".

Bear in mind that mixin was designed from scratch to be efficient when interacting with the expected lifecycle of the game and launchwrapper in particular, that's why there's no "discovery" phase where we scan for mixins, they're all predefined in configs; it's why all the heavy mixin pre-processing is done at the start of the phase rather than being calculated on the fly. The only other part of mixins which uses agents is a crude hack to make them hot-swappable, and it works okay provided you don't look at it too hard.

I'm willing to consider this, but only if there's actually a compelling use-case for it, which - as previously mentioned - "it would be cool" doesn't really qualify.

**Comment by me4502** [comment]
The main reason was to allow spigot plugins to use mixins

**Comment by Mumfrey** [comment]
> The main reason was to allow spigot plugins to use mixins

I'm not in favour of wasting time supporting deprecated platforms in all honesty.

**Comment by me4502** [comment]
Also SpongeVanilla, unless we provide a way for other things to hook tweaker

**Comment by Mumfrey** [comment]
SpongeVanilla uses LaunchWrapper, so can quite easily support additional tweakers. I'll provide the code if necessary.

**Comment by me4502** [comment]
So theoretically, if you got Spigot to run with LaunchWrapper, which kashike says is possible, you could get Mixins running?

**Comment by Mumfrey** [comment]
I wouldn't touch classloading in CB if you paid me a lot of money, but yes it's entirely possible. The main issue I see you hitting is that the classloader setup in CB is far from ideal with respect to integrating launchwrapper, but if someone were to get it working then yes mixin would work.

The main question at this point is **why**, and secondly I'm kind of annoyed that this has devolved into talking about getting mixin working with Spigot, which is **not** the original issue and means that either:

* you knew damn well I'd just close this straight away if you asked about spigot support, so were trying to be circumspect, which is frankly quite disingenuous
* you know that the original issue was a frivolous request and now you're clutching at straws to justify it

Either way, we're now out of scope for the original issue, so closing as invalid.

**Comment by me4502** [comment]
Yeah, I found that getting it running as a java agent was the best way of getting it running, so I made an issue about that. I didn't mention Spigot as I thought it was irrelevant, not that I thought you would close it straight away.

I always rework suggestions to try to expand the scope as far as possible, and limiting it to purely spigot, rather than the ability to provide runtime modification, would have limited the scope a fair bit.

As there is a workaround, I agree with having this closed if you don't see runtime mixins as a useful addition.
